### DISCUSSION TITLE
register may contain several contents

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#369] [Q&A]
The registers are designed to contain multiple contents. In most cases, this is not a problem, e.g., if the integer contained is a converted version of a numeric string (or vice versa). But sometimes they can contain independent contents. 
My question is this: Is there an indicator of what type the "last" result was?
I had a situation where a SAY statement didn't return the result of the last statement, but rather the content of a previous statement.

**Comment by adesutherland** [comment]
The compiler should be able to track this - is the issue you are having with compiled rexx or with hand-crafted rxas? If REXX then please post details - I will fix it.

However - you are right, and for level C we may need to do it dynamically.

Years ago (sigh!), I put this in rxvmintp.h:
```
typedef union {
    /* todo - these flag definitions are not used and are not correct */
    struct {
        unsigned int type_object : 1;
        unsigned int type_string : 1;
        unsigned int type_decimal : 1;
        unsigned int type_float : 1;
        unsigned int type_int : 1;
    };
    unsigned int all_type_flags;
} value_type;

struct value {
    /* bit field to store value status - these are explicitly set (not automatic at all) */
    value_type status;
   ...
```

So we have a status int / bit field for this. However, in the code above, the flag values are wrong. The compiler has its own definitions (in rxcpemit.c) which it used to track if arguments have values or not (for optional ones), or if the value of an argument needs to be kept the same or if the caller will discard the register anyway.

```
/* Register Type Flag Byte Values */
/* Used for optional arguments ONLY
 * set (1) means the register has a specified value */
#define REGTP_VAL 1

/* Used for "pass be value" large (strings, objects) registers ONLY
 * set (2) means that it is not a symbol so does not need copying as even if it is
 * changed the caller will not use its original value
 * Note: Small registers (int, float) are always copied as this is faster than
 *       setting and checking this flag anyway */
#define REGTP_NOTSYM 2
```

In sum, once we have start on level C we may need to use these flags consistently across the interpreter instructions and the compiler. However, at the moment they are ignored by the interpreter and the compiler can do what it likes with the flags.

A

**Reply by Peter-Jacob** [reply]
I noticed it in a handcrafted ASM. So I'd argue that registers should be cleared before they are reused. Assembler programmers should know what they are doing. Do we have an easy way (instruction) to clear a register?

**Reply by adesutherland** [reply]
There is the function value_zero() - we could have an instruction like "zero {reg}" to call that?
